### ISSUE TITLE
Quick function for converting bool array to frame indices

### ISSUE THREAD

**Issue by richardjgowers** [MDAnalysis/mdanalysis#1982]
We currently use  `np.arange(len(self))[frame]` where `frame` is a boolean array to extract frame indices in `Reader.__getitem__` and also [here](https://github.com/MDAnalysis/mdanalysis/pull/1934/files#r202064616) in @jbarnoud 's new work on iterating trajectories.

It would be nice to have a faster way of doing this!  Ie a cython function that takes the boolean array and spits out the frame indices without having to first generate the entire array of indices then slice it

**Comment by richardjgowers** [comment]
@ayushsuhane this isn't really related, but I think this should be an easy cython thing to write?

**Comment by ayushsuhane** [comment]
I tried this
```
from libcpp.vector cimport vector
cimport numpy as np
import numpy as np
import cython

@cython.boundscheck(False)
@cython.wraparound(False)
def bool2indices(np.ndarray[np.uint8_t, cast=True] boolvalues):
    cdef int i, size
    cdef vector[int] indices
    size = len(boolvalues)

    for i in range(size):
        if boolvalues[i]:
            indices.push_back(i)

    return np.asarray(indices)
```

but it seems like it is slower than the numpy indexing. i.e.
```
indices = np.arange(len(array))[array]
```
where, array is a bool array.

![image](https://user-images.githubusercontent.com/34154224/43800495-535d3e38-9a45-11e8-9de4-4b57ef21430a.png)

As far as I know,  memory views donot support handling of boolean array and we have to go to old buffers and casting uint8 to bool types. I am also looking for some other alternatives which can be advantageous. Any suggestions?

**Comment by zemanj** [comment]
`np.argwhere(array).ravel()` might do it...

**Comment by richardjgowers** [comment]
@zemanj "The output of argwhere is not suitable for indexing arrays. For this purpose use nonzero(a) instead" :p 

@ayushsuhane ok that's disappointing, but it looks like `np.nonzero` is an improvement

**Comment by ayushsuhane** [comment]
@richardjgowers  Yes it is, but not very significant.

![image](https://user-images.githubusercontent.com/34154224/43802274-2808de5e-9a4a-11e8-968f-880014a6ddba.png)

**Comment by kain88-de** [comment]
Your cython code is slow because it does a time of allocation. If you allocate the length of the vector beforehand and then fill it up your timings should be much better. 

So create a vector the same length as the Boolean input. Start to fill it up. Use the range of the vector you have filled with new values and copy it into a numpy array. This will reduce your allocation count to 2. 

Still why does np.where not work here?

**Comment by zemanj** [comment]
@kain88-de I once tried to speed up a C++ code by reducing the number of reallocs and wondered why this does exactly nothing.
It turned out that in reality, the number of allocations involving (expensive) syscalls were a lot fewer than expected: Whenever an allocation takes place, the system usually grants the calling process access to a little more memory than requested in order to avoid such consecutive syscalls. Now I'm really curious how that behaves in Cython!

**Comment by jbarnoud** [comment]
`np.where` seems to be the solution indeed.

**Comment by kain88-de** [comment]
[function with less allocations](https://gist.github.com/kain88-de/910aa5cbe1ed9ac2aba221cb1ea7473f)

Removing allocations makes the cython version faster. This naive implementation cannot beat `np.where` though.  BTW this would also work with a C++ vector class but compiling with the stdlib throughs errors for me right now and I don't want to fix that right now.

> @kain88-de I once tried to speed up a C++ code by reducing the number of reallocs and wondered why this does exactly nothing.

I haven't seen your code so I can't say much. But the larger allocation you have seen is likely not from the OS. Rather the vector type has a linear growth algorithm (python lists have that too!). For this the vector allocates to the next power of two and copies it's content to the new memory location. The new memory location is usually somewhere else, no guarantee is given by the language that it's just attached at the end. It now depends how your code was growing the array to see if it made a difference.  Also if the function you optimized was only a fraction of the total runtime the overall performance gain might be minimal. For C++ benchmarks I would also recommend using a framework like Google Benchmark for such optimizations. It's sometimes hard to judge the impact of an optimization.

**Comment by kain88-de** [comment]
The solution with the c++ vector is to call [`indices.reserve(size)`](https://en.cppreference.com/w/cpp/container/vector/reserve) before the loop to allocate all space that is needed.

**Comment by zemanj** [comment]
@kain88-de You're right, I mixed something up there. As you said, the padding added to allocations is certainly not done by the OS. To my limited knowledge, it should rather depend on the implementation of malloc/calloc/realloc (or whatever is used) in the installed standard library and should in principle not depend on the size of the data structure but rather be determined by memory alignment constraints. Thanks for mentioning the possibility of a linear growth algorithm, though, I'll look into that.
And thank you for pointing me to Google Benchmark, that sounds interesting! (So far I used gprof where applicable.)
